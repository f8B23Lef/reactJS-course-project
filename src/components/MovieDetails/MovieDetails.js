import React from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSearch } from '@fortawesome/free-solid-svg-icons';
import Logo from '../Logo/Logo';
import { getYearFromDate, getHoursAndMinutes } from '../../utils/utils';
import './MovieDetails.scss';

export default function MovieDetails(props) {
  const {
    movie,
    setShowMovieDetails,
  } = props;

  return (
    <div className='movie-details'>
      <div className='movie-details__header'>
        <Logo />
        <span
          className='search__icon'
          onClick={() => setShowMovieDetails(false)}
        >
          <FontAwesomeIcon icon={faSearch} />
        </span>
      </div>
      <div className='movie-details__content'>
        <img src={movie.poster} alt='movie poster' />
        <div className='info'>
          <div className='title'>
            {movie.title}
            <span className='rating'>{movie.rating}</span>
          </div>
          <div className='genre'>{movie.genres.join(', ')}</div>
          <div className='duration'>
            <span>{getYearFromDate(movie.releaseDate)}</span>
            <span className='runtime'>{getHoursAndMinutes(movie.runtime)}</span>
          </div>
          <div className='description'>{movie.description}</div>
        </div>
      </div>
    </div>
  );
}

MovieDetails.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string,
    releaseDate: PropTypes.string,
    url: PropTypes.string,
    rating: PropTypes.number,
    genres: PropTypes.arrayOf(PropTypes.string),
    runtime: PropTypes.number,
    description: PropTypes.string,
    poster: PropTypes.string.isRequired,
  }),
  setShowMovieDetails: PropTypes.func.isRequired,
};

MovieDetails.defaultProps = {
  movie: {
    title: '',
    releaseDate: '',
    url: '',
    rating: 0,
    genres: [],
    runtime: 0,
    description: '',
  },
};
