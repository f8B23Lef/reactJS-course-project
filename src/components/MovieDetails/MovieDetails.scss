@import '../../styles/colors';

.movie-details {
  padding: 20px 35px;

  &__header {
    display: flex;
    justify-content: space-between;

    .search__icon {
      font-size: 18px;
      color: $red;
      cursor: pointer;
    }
  }

  &__content {
    display: flex;
    margin-top: 20px;

    .info > div {
      padding-bottom: 10px;
    }

    img {
      width: 200px;
      height: 300px;
      margin-right: 40px;
    }

    .title {
      font-size: 18px;
      text-transform: uppercase;
    }

    .rating {
      margin-left: 10px;
      padding: 5px;
      border: 1px solid $white;
      border-radius: 50%;
    }

    .genre {
      opacity: 0.6;
    }

    .duration {
      color: $red;
      font-size: 18px;
    }
    
    .runtime {
      margin-left: 10px;
    }

    .description {
      font-size: 16px;
      opacity: 0.6;
    }
  }
}
