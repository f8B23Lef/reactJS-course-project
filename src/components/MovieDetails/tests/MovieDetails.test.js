import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { screen, render, fireEvent } from '@testing-library/react';
import MovieDetails from '../MovieDetails';

jest.mock('../../Logo/Logo', () => () => 'SomeComponent');

const mockedUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

describe('MovieDetails', () => {
  const movie = {
    id: 1,
    title: 'Movie 1',
    release_date: '2022-10-01',
    vote_average: 8,
    genres: ['crime', 'action'],
    runtime: 120,
    overview: 'Some overview',
    poster_path: 'www.movie1.com',
  };
  const setShowMovieDetails = jest.fn();

  it('renders MovieDetails snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <MovieDetails movie={movie} setShowMovieDetails={setShowMovieDetails} />
      </BrowserRouter>,
    );

    expect(asFragment(
      <MovieDetails movie={movie} setShowMovieDetails={setShowMovieDetails} />,
    )).toMatchSnapshot();
  });

  it('finds movie title', () => {
    render(
      <BrowserRouter>
        <MovieDetails movie={movie} setShowMovieDetails={setShowMovieDetails} />
      </BrowserRouter>,
    );

    expect(screen.getByText(movie.title)).toBeInTheDocument();
  });

  it('calls setShowMovieDetails() on Search icon click', () => {
    render(
      <BrowserRouter>
        <MovieDetails movie={movie} setShowMovieDetails={setShowMovieDetails} />
      </BrowserRouter>,
    );

    const icon = screen.getByTestId('search-icon');
    fireEvent.click(icon);

    expect(setShowMovieDetails).toHaveBeenCalledWith(false);
  });

  it('calls navigate() on Search icon click', () => {
    render(
      <BrowserRouter>
        <MovieDetails movie={movie} setShowMovieDetails={setShowMovieDetails} />
      </BrowserRouter>,
    );

    const icon = screen.getByTestId('search-icon');
    fireEvent.click(icon);

    expect(mockedUseNavigate).toHaveBeenCalledWith('/');
  });
});
