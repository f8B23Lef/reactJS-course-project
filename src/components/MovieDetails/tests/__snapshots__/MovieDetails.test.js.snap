// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails renders MovieDetails snapshot 1`] = `
<DocumentFragment>
  <div
    class="movie-details"
  >
    <div
      class="movie-details__header"
    >
      SomeComponent
      <span
        class="search__icon"
        data-testid="search-icon"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-magnifying-glass "
          data-icon="magnifying-glass"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <div
      class="movie-details__content"
    >
      <img
        alt="movie poster"
        src="www.movie1.com"
      />
      <div
        class="info"
      >
        <div
          class="title"
        >
          Movie 1
          <span
            class="rating"
          >
            8
          </span>
        </div>
        <div
          class="genre"
        >
          crime, action
        </div>
        <div
          class="duration"
        >
          <span>
            2022
          </span>
          <span
            class="runtime"
          >
            2h 0min
          </span>
        </div>
        <div
          class="description"
        >
          Some overview
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
